src/music/Instruments.ts

```diff
@@ -0,0 +1,90 @@
+import Instrument from "./Instrument.js";
+import { Envelopes, AdsrConfig } from "./Envelopes.js";
+import { Note, Notes } from "./Notes.js";
+import { SampleData } from "./Samples.js";
+
+type OscType = typeof OscillatorNode.prototype.type;
+
+export interface OscillatorConfig {
+    type: OscType; // either 'sine', 'triangle', 'square', or 'sawtooth'
+    detune?: number; // detune in cents
+}
+
+export class AdsrOscillatorInstrument extends Instrument {
+
+    type: OscType;
+    _detune: number;
+    env: AdsrConfig;
+
+    constructor(osc: OscillatorConfig, env: AdsrConfig, volume: number = 1) {
+        super(volume);
+        this.type = osc.type;
+        this._detune = 1;
+        this.detune = osc.detune || 0;
+        this.env = env;
+    }
+
+    set detune(n: number) {
+        this._detune = Math.pow(2, n / 1200);
+    }
+
+    scheduleNote(context: AudioContext, note: Note): AudioNode {
+        const freqs: number[] = Notes.detuneWithCoeff(note.note, this._detune);
+        // note ending frequencies are either where they started or at the endNote
+        const endfreqs: number[] = (note.endNote === undefined) ? freqs : Notes.detuneWithCoeff(note.endNote, this._detune);
+        // create envelope
+        const gain: GainNode = Envelopes.createAdsrEnvelope(context, note.start, note.duration,
+            this.env, Math.min(1 / freqs.length, 1) * this.volume);
+        const end = note.start + note.duration + (this.env.sustain || 0);
+        // initialize oscillator(s)
+        const oscs: OscillatorNode[] = freqs.map((freq, i) => {
+            const osc = context.createOscillator();
+            osc.type = this.type;
+            osc.frequency.setValueAtTime(freq, note.start);
+            osc.frequency.linearRampToValueAtTime(endfreqs[i], end);
+            osc.start(note.start);
+            osc.stop(end);
+            osc.connect(gain);
+            return osc;
+        });
+        oscs[0].onended = () => {
+            gain.disconnect();
+        };
```

Comment:
I don't have enough context to know why this is only attached to the first node

Reply:
From an in-person discussion, the `GainNode` should be disconnected when the last `OscillatorNode` is done. Theoretically, they should all end at the same time, so it's probably fine to only attach it to the first one. A cleaner way to do this would probably be to promisify all of the `OscillatorNode` `onended` callbacks, then wrap the array in `Promise.all` and detach after all of them have finished.